### ISSUE TITLE
syntax highlighting of code blocks only works for default theme

### ISSUE THREAD

**Issue by nylki** [johnfactotum/foliate#1315]
**Describe the bug**
Code blocks (monospace blocks with syntax highlighting) don't work for most themes, only the default theme.

**To Reproduce**
1. open an ePub that has a code block with syntax highlighting
2. Click on "Fonts and Layout" (settings top right)
3. click on "Colors" (tab bar item)
4. Select a different color theme (other than default)
5. code block is not syntax highlighted anymore, using the same font color as the regular text

**Expected behavior**
When changing the color theme I expect syntax highlight of code blocks to still be visible.

**Screenshots**
default theme, featuring syntax highlighting and different bg-color:
![Bildschirmfoto vom 2024-05-14 20-18-47](https://github.com/johnfactotum/foliate/assets/1710598/6542c363-14c3-4f9a-aefa-ed932d146212)

solarized (or sepia) theme without syntax highlighting and same bg-color:
![Bildschirmfoto vom 2024-05-14 20-19-08](https://github.com/johnfactotum/foliate/assets/1710598/dcdb0eb3-6dca-47df-be18-05552f28d280)

**Version:**
 - Foliate version:
 - OS/Distribution and version: Fedora Linux 40 (Workstation Edition) 
 - Desktop environment: Gnome 46
 - Installation method: Official Fedora repo RPM (but also checked on the flathub version)

**Comment by johnfactotum** [comment]
This is a feature to ensure that the text is legible. I *think* this is also the behavior in Apple Books and Readium/Thorium.

I suppose the easiest way out would be to just add an option to disable this.

**Comment by nylki** [comment]
> This is a feature to ensure that the text is legible. I _think_ this is also the behavior in Apple Books and Readium/Thorium.
> 
> I suppose the easiest way out would be to just add an option to disable this.

Thank you for your quick response! Just to avoid confusion: Is my assumption correct that the non-default themes don't have a code-block specific stylesheet that would fit their color scheme? (so they are colored as if they are plaintext).

**Comment by johnfactotum** [comment]
The syntax highlighting is part of the stylesheet of the EPUB file. Most EPUB files are designed to be displayed on a white background only and look broken when you apply a custom theme. So when using non-white background, Foliate overrides all foreground and background colors so that the theme would be applied consistently and the text would remain readable. See 29eed7e6dbd114cc095aee3efe93ece24a4a399c.

**Comment by nylki** [comment]
@johnfactotum Thanks for explaining, makes way more sense now :slightly_smiling_face: ! I did not know that the syntax coloring is baked into the EPUB itself.
I was hoping that code blocks were  somehow defined in a more generic way in EPUB (like a `<code></code>`-block), so that applying a syntax theme would be relatively easy.

I wonder if it would be feasible to heuristically identify code blocks and then define a custom theme for those section.

Anyways, feel free to close the issue if you like.

**Comment by johnfactotum** [comment]
Yes, the code should be marked up with the `<code>` element. However, to quote the [HTML Standard](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-code-element):

> There is no formal way to indicate the language of computer code being marked up.

So automatically applying syntax highlighting would be problematic.

**Comment by nylki** [comment]
I see. So, just to entertain the idea: although there is no formal way, auto-detection of the language inside a block could be an option (perhaps as an option that is disabled by default, or more involved, via a button to change the detected language). For example highlight.js does [supports auto-detection](https://github.com/highlightjs/highlight.js/?tab=readme-ov-file#in-the-browser) of languages.

**Comment by johnfactotum** [comment]
Yes, that would be possible in theory, but probably too localized and complicated to be ever implemented in Foliate itself.

Some alternatives, when the book already has syntax highlighting, which most of them do, are
- Allow disabling color overrides
- Use other methods such as filters to change the colors
- Bring back support for `__ibooks_internal_theme`, and then the book itself can use `-webkit-text-fill-color` to force the text color

**Comment by Jopp-gh** [comment]
I had a similar styling issue with a theme and a possible solution could be to make:
- syntax colors slightly brighter on dark themes 
- syntax colors slightly darker on light themes/ keep the original colors

I don't know what color coordinate system Foliate uses, but HSL would be fine.

**Comment by Atique29** [comment]
Enabling "Invert Colors in Dark Mode" option brought back the syntax highlight that went missing upon selecting a non-default theme.
